trace_decoder/src/compact/compact_to_partial_trie.rs

```diff
@@ -8,183 +11,261 @@ use mpt_trie::{
 
-    create_partial_trie_from_compact_node(remaining_node)
-}
+    fn process_empty(&self) -> CompactParsingResult<()> {
+        // Nothing to do.
+        Ok(())
+    }
 
-pub(super) fn create_partial_trie_from_compact_node(
-    node: NodeEntry,
-) -> CompactParsingResult<CompactToPartialTrieExtractionOutput> {
-    let mut output = CompactToPartialTrieExtractionOutput::default();
+    fn process_hash(&mut self, curr_key: Nibbles, hash: TrieRootHash) -> CompactParsingResult<()> {
+        self.get_trie().insert(curr_key, hash);
 
-    create_partial_trie_from_compact_node_rec(Nibbles::default(), &node, &mut output)?;
+        Ok(())
+    }
 
-    Ok(output)
-}
+    fn process_leaf(
+        &mut self,
+        curr_key: Nibbles,
+        leaf_key: &Nibbles,
+        leaf_node_data: &LeafNodeData,
+    ) -> CompactParsingResult<()>;
 
-// TODO: Consider putting in some asserts that invalid nodes are not appearing
-// in the wrong trie type (eg. account )
-pub(super) fn create_partial_trie_from_compact_node_rec(
-    curr_key: Nibbles,
-    curr_node: &NodeEntry,
-    output: &mut CompactToPartialTrieExtractionOutput,
-) -> CompactParsingResult<()> {
-    trace!("Processing node {} into `PartialTrie` node...", curr_node);
+    fn process_extension(
+        &mut self,
+        curr_key: Nibbles,
+        ext_node_key: &Nibbles,
+        ext_child: &NodeEntry,
+    ) -> CompactParsingResult<()> {
+        let new_k = curr_key.merge_nibbles(ext_node_key);
+        create_partial_trie_from_compact_node_rec(new_k, ext_child, self)?;
 
-    match curr_node {
-        NodeEntry::Branch(n) => process_branch(curr_key, n, output),
-        NodeEntry::Code(c_bytes) => process_code(c_bytes.clone(), output),
-        NodeEntry::Empty => process_empty(),
-        NodeEntry::Hash(h) => process_hash(curr_key, *h, &mut output.trie),
-        NodeEntry::Leaf(k, v) => process_leaf(curr_key, k, v, output),
-        NodeEntry::Extension(k, c) => process_extension(curr_key, k, c, output),
+        Ok(())
     }
+
+    fn get_trie(&mut self) -> &mut HashedPartialTrie;
 }
 
-fn process_branch(
-    curr_key: Nibbles,
-    branch: &[Option<Box<NodeEntry>>],
-    output: &mut CompactToPartialTrieExtractionOutput,
-) -> CompactParsingResult<()> {
-    for (i, slot) in branch.iter().enumerate().take(16) {
-        if let Some(child) = slot {
-            // TODO: Seriously update `mpt_trie` to have a better API...
-            let mut new_k = curr_key;
-            new_k.push_nibble_back(i as Nibble);
-            create_partial_trie_from_compact_node_rec(new_k, child, output)?;
+#[derive(Debug)]
+pub(super) enum UnexpectedCompactNodeType {
+    AccountLeaf,
+    Code,
+}
+
+impl Display for UnexpectedCompactNodeType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            UnexpectedCompactNodeType::AccountLeaf => write!(f, "AccountLeaf"),
+            UnexpectedCompactNodeType::Code => write!(f, "Code"),
         }
     }
+}
```

Comment:
Unless we know we want a `Display` impl I'd refrain from implementing it. The impl here is essentially identical to the `Debug` impl, so why bother? I think `Display` impls can wait until there's a clear need for more advanced printing needs.

Reply:
It does appear in a variant for `CompactParsingError`. We *could* use the `Debug` impl, but it would instead print out `UnexpectedCompactNodeType::AccountLeaf` instead of just `AccountLeaf`. I think it's worth using `Display` just for that.

Reply:
I take that back. I just tested this, and it does print out `AccountLeaf`.